### PR TITLE
Resolve: "ADD: BCH legacy address validation"

### LINKED ISSUE #335: ADD: BCH legacy address validation
# current

* BCH legacy addresses are interpreted as invalid and not allowed to send to
* Users need to use services like https://tools.bitcoin.com/cash-address-converter/ to convert

# Desired
* convert all BCH addresses to cashAddr
* to a validation check *after* converting
* send always to cashAddre (ie, convert before sending)

Use this method here:
https://github.com/jbdtky/bitcoincashjs/blob/master/src/address.js#L348

Or just import and use:

```
const bch = require('bitcoincashjs');

const Address = bch.Address;
const CashAddrFormat = Address.CashAddrFormat;

const address = new Address('1MF7A5H2nHYYJMieouip2SkZiFZMBKqSZe');

console.log(address.toString()) // 1MF7A5H2nHYYJMieouip2SkZiFZMBKqSZe
console.log(address.toString(CashAddrFormat)) // bitcoincash:qr0q67nsn66cf3klfufttr0vuswh3w5nt5jqpp20t9
```

### PR DESCRIPTION
**Pull request by mikewyszinski**
Fix #335

### CODE REVIEW

File: packages/xchain-bitcoincash/src/utils.ts [hunk 1]
```diff
@@ -192,9 +192,10 @@ export const parseTransaction = (tx: Transaction): Tx => {
  */
 export const validateAddress = (address: string, network: Network): boolean => {
   try {
-    bitcash.address.toOutputScript(toLegacyAddress(address), bchNetwork(network))
-    return true
+    const toAddress = toCashAddress(address)
+    return bchaddr.isValidAddress(toAddress) && bchaddr.detectAddressNetwork(toAddress) === network
   } catch (error) {
+    console.log(error)
```
Comment (veado) [inline packages/xchain-bitcoincash/src/utils.ts:198]: Still needed ^? Or maybe `console.warn`?

File: packages/xchain-bitcoincash/src/client.ts [hunk 2]
```diff
@@ -437,6 +437,7 @@ class Client implements BitcoinCashClient, XChainClient {
       const keyPair = this.getBCHKeys(this.phrase, derivationPath)
 
       utxos.forEach((utxo, index) => {
+        console.log(JSON.stringify(utxo, null, ''))
```
Comment (veado) [inline packages/xchain-bitcoincash/src/client.ts:440]: Still needed ^?

File: packages/xchain-bitcoincash/src/utils.ts [hunk 3]
```diff
@@ -192,9 +192,10 @@ export const parseTransaction = (tx: Transaction): Tx => {
  */
 export const validateAddress = (address: string, network: Network): boolean => {
   try {
-    bitcash.address.toOutputScript(toLegacyAddress(address), bchNetwork(network))
-    return true
+    const toAddress = toCashAddress(address)
+    return bchaddr.isValidAddress(toAddress) && bchaddr.detectAddressNetwork(toAddress) === network
```
Comment (veado) [inline packages/xchain-bitcoincash/src/utils.ts:196]: Comparing `Network` (xchain) with `Network` (bchaddrjs) is currently working, but still unsafe. A simple helper might help
```ts
const toBCHAddressNetwork = (network: Network /* xchain */): bchaddr.Network => network === ' testnet' ? bchaddr.Network.Testnet : bchaddr.Network.Mainnet
```
`bchaddr.Network` is defined here https://github.com/ealmansi/bchaddrjs/blob/f3f4664b2fe5d9edba253dcc7c569b0a5e93002c/src/bchaddr.js#L39

**Review by veado — CHANGES_REQUESTED** [review]
LGTM, just few things

File: packages/xchain-bitcoincash/src/utils.ts (hunk 1 above)
Comment (mikewyszinski) [reply packages/xchain-bitcoincash/src/utils.ts:198]: yeah good catch...i think i'm going to remove the catch altogether..

File: packages/xchain-bitcoincash/src/client.ts (hunk 2 above)
Comment (mikewyszinski) [reply packages/xchain-bitcoincash/src/client.ts:440]: damn thx! sloppy debugging left over by me

File: packages/xchain-bitcoincash/src/utils.ts (hunk 3 above)
Comment (mikewyszinski) [reply packages/xchain-bitcoincash/src/utils.ts:196]: fixed :-)

Comment (veado) [reply packages/xchain-bitcoincash/src/utils.ts]: I was wrong to assume a type `bchaddr.Network` exists. e17339b added a `toBCHAddressNetwork` helper.

**Pull request merged**